.travis.yml

```diff
@@ -37,7 +37,9 @@ jobs:
       before_install:
         - mkdir bin ; ln -s /usr/bin/gcc-5 bin/gcc
       # env: COMPILER=g++-5 SAN_FLAGS="-fsanitize=undefined -fno-sanitize-recover -fno-omit-frame-pointer"
-      env: COMPILER="ccache g++-5"
+      env:
+        - COMPILER="ccache g++-5"
+        - EXTRA_CXXFLAGS="-DNDEBUG"
```

Comment:
My feeling is this should be a separate build since in general I don't think we've replaced all asserts with invariant (@peterschrammel may have an opinion on this).

Reply:
I didn't think a separate build was needed, since all assertions are checked in the clang build. Thus an escape would be an assertion that only fails with gcc, but not with clang. That scenario seemed sufficiently unlikely to me.

Reply:
The  problem is that this build is run in the first stage, which should provide quick diagnostic information. A segfault there wouldn't provide useful information. Once all asserts have been replaced this problem would disappear, of course.

I am not quite sure what is the motivation for this change? Have you observed different behaviour w/wo NDEBUG?

Reply:
The motivation for the change is to ensure that compiling with -DNDEBUG does not get broken again. So there should be a build in the CI that checks this. That build could either be a new one, or the modification of an existing one. The PR does the latter, @thk123 suggested the former.

Reply:
@andreast271 The issue isn't so much checking as it's things with side effects in asserts and / or INVARIANT et al.  I'm not sure I'm so confident that you won't find difference between compilers; but maybe I'm just old an cynical.

I'm generally +1 super enthusiastic about -DNDEBUG not breaking things and testing it.  Whether adding to an existing build or a new one; I'm not sure I care that much.

Can I request that we also have -DCPROVER_INVARIANT_DO_NOT_CHECK to prevent the "next generation" version of this problem as well.

Reply:
Building with -DCPROVER_INVARIANT_DO_NOT_CHECK is already broken. I have seen 6 different errors:
```
./std_types.h:143:3: error: must specify at least one argument for '...' parameter of variadic macro [-Werror,-Wgnu-zero-variadic-macro-arguments]

goto_symex_state.cpp:532:40: error: 'nullptr_exceptiont' does not refer to a
      value
  INVARIANT_STRUCTURED(dirty!=nullptr, nullptr_exceptiont, "dirty is null");

expr_util.cpp:37:16: error: unused variable 'type' [-Werror,-Wunused-variable]
  const typet &type=expr.type();

c_types.cpp:72:1: error: control may reach end of non-void function
      [-Werror,-Wreturn-type]

flattening/flatten_byte_operators.cpp:202:11: error: variable 'little_endian' is
      used uninitialized whenever 'if' condition is false
      [-Werror,-Wsometimes-uninitialized]
  else if(src.id()==ID_byte_extract_big_endian)
}

refinement/string_refinement.cpp:2076:13: error: unused function
      'is_valid_string_constraint' [-Werror,-Wunused-function]
```
I will suggest fixes for the first two errors in the next commit, and I'd like to hear suggestions about the other  issues. Meanwhile in the travis build I will add -DCPROVER_INVARIANT_DO_NOT_CHECK and disable the specific warnings that cause compilation failures. The warnings can be re-enabled once the macros are fixed.

Reply:
Suggestions for issues 3-6: 
3. `const typet &type=expr.type();` -> just use `expr.type()` in those `PRECONDITION` checks.
4. and 5.: Make the last `else if` an `else` and move the condition tested into the `INVARIANT`.
6. Surround the declaration and definition of `is_valid_string_constraint` by `#ifndef NDEBUG`.